dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStore.java

```diff
@@ -562,44 +576,48 @@ private String getFromSubQuery( TrackedEntityInstanceQueryParams params, boolean
-            .append( "TEI.lastupdated, " )
-            .append( "TEI.inactive, " )
-            .append( "TEI.trackedentitytypeid, " )
-            .append( "TEI.potentialduplicate, " )
-            .append( "TEI.deleted, " )
-            .append( "OU.uid as ou, " )
-            .append( "OU.name as ouname " )
-            .append( getFromSubQueryOrderAttributes( params ) )
-            .toString();
+        LinkedHashSet<String> columns = new LinkedHashSet<>(
+            List.of( "TEI.trackedentityinstanceid", "TEI.uid", "TEI.created", "TEI.lastupdated",
+                "TEI.inactive", "TEI.trackedentitytypeid", "TEI.potentialduplicate", "TEI.deleted", "OU.uid as ou",
+                "OU.name as ouname " ) );
+
+        for ( OrderParam orderParam : params.getOrders() )
+        {
+            Optional<TrackedEntityInstanceQueryParams.OrderColumn> orderColumn = findColumn( orderParam.getField() );
+
+            if ( orderColumn.isPresent() )
+            {
+                columns.add( orderColumn.get().getColumnWithTableAliasSql() );
+            }
+            else
+            {
+                if ( sortableAttributesAndFilters( params ).stream()
+                    .anyMatch( i -> i.getItem().getUid().equals( orderParam.getField() ) ) )
+                {
+                    columns.add( statementBuilder.columnQuote( orderParam.getField() ) + ".value AS " +
+                        statementBuilder.columnQuote( orderParam.getField() ) );
+                }
+            }
+        }
+
+        return "SELECT " + Strings.join( ", ", columns );
     }
 
     /**
-     * Generates a list of columns for projections if we are ordering by
-     * attributes.
+     * Get a list of QueryItem that contains sortable attributes also defined as
+     * filers
      *
      * @param params
-     * @return a list of columns to be used in the subquery SQL projection, or
-     *         empty string if no attributes are used for ordering.
+     * @return List of QueryItem
      */
-    private String getFromSubQueryOrderAttributes( TrackedEntityInstanceQueryParams params )
+    private List<QueryItem> sortableAttributesAndFilters( TrackedEntityInstanceQueryParams params )
     {
-        StringBuilder orderAttributes = new StringBuilder();
-
-        for ( QueryItem orderAttribute : getOrderAttributes( params ) )
-        {
-            orderAttributes
-                .append( ", " )
-                .append( statementBuilder.columnQuote( orderAttribute.getItemId() ) )
-                .append( ".value AS " )
-                .append( statementBuilder.columnQuote( orderAttribute.getItemId() ) );
-        }
-
-        return orderAttributes.toString();
+        List<String> ordersIdentifier = params.getOrders().stream()
+            .map( OrderParam::getField )
+            .collect( Collectors.toList() );
+        return params.getAttributesAndFilters().stream()
+            .filter( queryItem -> ordersIdentifier.contains( queryItem.getItemId() ) )
```

Comment:
I would rather use a Set then a List (even if ordersIdentifier might not contain large N items)